### PR TITLE
Single psf optimizer

### PR DESCRIPTION
**Pull request by rgiordan**
This makes some additional optimizations from fit_psf_in_celeste.

### CODE REVIEW

**Comment by jeff-regier** [comment]
@kbarbary Is there a workaround to get FITSIO running on travis while
heasarc.gsfc.nasa.gov is down?

**Comment by kbarbary** [comment]
Yes; install libcfitsio3 with apt-get on travis. This is even [what the FITSIO.jl travis tests](https://github.com/JuliaAstro/FITSIO.jl/blob/master/.travis.yml) do.

**Comment by jeff-regier** [comment]
Jula 0.4 tests pass now. We're getting a new error message with Julia 0.5. It doesn't even give us a line number, just somewhere in `Types.jl`. Looks like a compile time error. Did Julia's iteration syntax change recently?

```
ERROR: LoadError: LoadError: LoadError: syntax: unhandled expr (error invalid iteration specification)
 in include(::ASCIIString) at ./boot.jl:240
 in include_from_node1(::ASCIIString) at ./loading.jl:417
 in include(::ASCIIString) at ./boot.jl:240
 in include_from_node1(::ASCIIString) at ./loading.jl:417
 in eval(::Module, ::Any) at ./boot.jl:243
 [inlined code] from ./sysimg.jl:11
 in require(::Symbol) at ./loading.jl:348
 in include(::ASCIIString) at ./boot.jl:240
 in include_from_node1(::UTF8String) at ./loading.jl:417
 in process_options(::Base.JLOptions) at ./client.jl:262
 in _start() at ./client.jl:318
while loading /home/travis/.julia/v0.5/Celeste/src/Types.jl, in expression starting on line 1
while loading /home/travis/.julia/v0.5/Celeste/src/Celeste.jl, in expression starting on line 5
while loading /home/travis/.julia/v0.5/Celeste/test/runtests.jl, in expression starting on line 3
```

**Comment by jeff-regier** [comment]
With the NERSC version of Julia 0.5 `using Celeste` succeeds, so I suspect iteration syntax has changed in Julia nightly.

With the NERSC version of Julia 0.5 I get error messages about `CelesteTypes` not found. Unfortunately I probably won't get to look into this more today.

File: src/Types.jl [hunk 1]
```diff
@@ -483,6 +484,23 @@ getids(::Type{UnconstrainedParams}) = ids_free
 length(::Type{UnconstrainedParams}) =  6 + 2*Ia + 2*(B-1)*Ia + (D-1)*Ia + Ia-1
 
 
+# Parameters for a representation of the PSF
+type PsfParams <: Types.ParamSet
```
Comment (kbarbary) [inline src/Types.jl]: I think this is the problem. This is already in the `Types` module, so this should be

``` julia
type PsfParams <: ParamSet
```

no?

**Comment by rgiordan** [comment]
Ok, let's try without the Types prefix.

**Comment by jeff-regier** [comment]
I wonder how `test_elbo_values.jl` passed with Julia 0.4 (but not 0.5)...it was calling `CelesteTypes` rather than `Types`.

**Comment by jeff-regier** [comment]
This works on nersc's version of julia 0.5-dev now, but on travis (and with the current nightly of julia), `using Celeste` fails still. I guess the extra `Types` wasn't the problem. Do you we want to just remove Julia nightly from the travis file, now that nightly is deviating from the version of julia 0.5 on nersc?

**Comment by rgiordan** [comment]
How should I install or use libcfitsio3 on Cori?  I'm getting this same
error when I try to install the Celeste package for the first time.

**Comment by jeff-regier** [comment]
Can you copy the whole

/global/homes/j/jregier/.julia/v0.4/FITSIO

over to your home directory? I'll check the permissions.

**Comment by rgiordan** [comment]
I'm currently getting permission denied.

**Comment by rgiordan** [comment]
Looks like the problem is with your home directory itself:

rgiordan@cori08:~> ls -la /global/homes/j/jregier/
ls: cannot open directory /global/homes/j/jregier/: Permission denied
rgiordan@cori08:~> ls -ld /global/homes/j/jregier/
drwx--x--x 11 jregier jregier 4096 Mar 20 15:30 /global/homes/j/jregier/

**Comment by jeff-regier** [comment]
Hmmm...can you copy it over from your laptop? (scp -r)

**Comment by rgiordan** [comment]
I can try.  Is there a reason you can't make your home directory readable?
Kyle's is (but his .julia directory is not).

**Comment by jeff-regier** [comment]
I thought mine was. Hang on

**Comment by jeff-regier** [comment]
How about now?

**Comment by rgiordan** [comment]
Not all the subdirectories are executable, e.g.:

rgiordan@cori08:~> ls -ltd /global/homes/j/jregier/.julia/v0.4/Logging/
drwxr-xr-- 5 jregier jregier 512 Mar 16 16:59
/global/homes/j/jregier/.julia/v0.4/Logging

**Comment by rgiordan** [comment]
scp is taking a really long time, it will be helpful if you can fix the
permissions.

**Comment by jeff-regier** [comment]
Yeah just copy the FITSIO directory, then run Pkg("resolve").

**Comment by rgiordan** [comment]
Could you fix the permissions please?

rgiordan@cori08:~> cp -R /global/homes/j/jregier/.julia/v0.4/FITSIO/
.julia/v0.4/
cp: cannot stat `/global/homes/j/jregier/.julia/v0.4/FITSIO/.git/refs':
Permission denied

**Comment by jeff-regier** [comment]
I've run

chmod -R a+r ~/.julia

Does that work?

**Comment by rgiordan** [comment]
No, because you have not made the subdirectories executable as well.

**Comment by rcthomas** [comment]
You need the directories to be made executable.  `chmod -R a+rX` I believe.

But this is bizarre, how come Ryan can't build FITSIO but others can...?!

**Comment by rgiordan** [comment]
The nasa website probably wasn't down when you guys installed Celeste.

**Comment by rcthomas** [comment]
Damn budget cuts...

**Comment by rgiordan** [comment]
So when I use my local copy of FITSIO, if I run Pkg.build() on it it tries
to hit the nasa server, and if I try to use it I get the following error:

julia> using FITSIO
INFO: Precompiling module FITSIO...
ERROR: LoadError: LoadError: LoadError: Unable to load

libcfitsio (/usr/lib/x86_64-linux-gnu/libcfitsio.so)

Please re-run Pkg.build(package), and restart Julia.

**Comment by jeff-regier** [comment]
I know, right.

This is running:
jregier@cori10~/.julia $  chmod -R a+rX ~/.julia

**Comment by rgiordan** [comment]
Sorry, by "local copy" I meant the version from my laptop.

Sorry guys, Dora's flight is landing shortly, so I've got to sign off for a
while, probably most of the rest of the evening.

**Comment by jeff-regier** [comment]
Your version of FITSIO is probably only partial installed, that's why it's
trying to download it when you build. Please let me know if still can't get
what's in my directory...I can't test this out myself unfortunately wo
another nersc account.

**Comment by rgiordan** [comment]
The copy from your account worked, Jeff!

**Comment by jeff-regier** [comment]
Glad to hear it  :-)

**Comment by kbarbary** [comment]
There's probably a nersc module that provides libcfitsio. That should solve
this in a more dependable way.

**Comment by rcthomas** [comment]
Confused --- thought your Julia module built CFITSIO from scratch (looked like it did).  Is that optional?  If so then what you want is `module load cfitsio` --- this provides CFITSIO headers and libraries.  This would need to be done before you launch Julia to build FITSIO.

**Comment by kbarbary** [comment]
Yes, its optional. When you build FITSIO, it first tries to load libcfitsio to see if it is already available somewhere. If that fails, then it builds from source. At the end of building, it saves a record of the path to the library it found by writing a stub file `deps/deps.jl` which then gets used by FITSIO.

**Comment by jeff-regier** [comment]
Huzzah!

**Comment by jeff-regier** [comment]
@kpamnany   @rcthomas  The master branch is running fast now, with Ryan's PSF work merged. You two should be all set now to kick off large runs. Please file bugs as you encounter them....the "access to undefined reference" seems to be pretty common.

**Comment by rgiordan** [comment]
It would be very useful to me if bugs could come with an object id, run,
camcol, and field so that they're easy to reproduce in isolation.  I've got
to do some other work for a couple days, but if the next bug reporter feels
like adding a --verbose flag to infer() to report this kind of information
it would be very helpful.

**Comment by kpamnany** [comment]
Working on this.

**Comment by kpamnany** [comment]
Sadly, Cori is down today.

Can I get appropriate parameters for a single node run as well as something appropriate for 8 nodes, etc.? That is, what is the rad/dec range I should use?

**Comment by jeff-regier** [comment]
Hi Kiran, test_infer.jl shows arguments that work, but it only processes 4
sources. You could make the bounding box bigger to include enough work for
multiple threads.

**Comment by kpamnany** [comment]
Is there something that specifically limits the number of sources to 4, or it just so happens that that bounding box has only 4 sources?

**Comment by jeff-regier** [comment]
The latter, there just happen to be 4 sources in that box.

**Comment by kpamnany** [comment]
No results yet. Cori is back up, but problems with the filesystem I think -- can't run anything. :(

**Comment by jeff-regier** [comment]
It's working for me. Have you tried logging out and then reconnecting?

**Pull request merged**